conan/tools/env/environment.py

```diff
@@ -0,0 +1,291 @@
+import fnmatch
+import os
+import textwrap
+from collections import OrderedDict
+
+from conans.errors import ConanException
+from conans.util.files import save
+
+
+class _EnvVarPlaceHolder:
+    pass
+
+
+class _Sep(str):
+    pass
+
+
+class _PathSep:
+    pass
+
+
+def environment_wrap_command(filename, cmd, cwd=None):
+    assert filename
+    filenames = [filename] if not isinstance(filename, list) else filename
+    bats, shs = [], []
+    for f in filenames:
+        full_path = os.path.join(cwd, f) if cwd else f
+        if os.path.isfile("{}.bat".format(full_path)):
+            bats.append("{}.bat".format(f))
+        elif os.path.isfile("{}.sh".format(full_path)):
+            shs.append("{}.sh".format(f))
+    if bats and shs:
+        raise ConanException("Cannot wrap command with different envs, {} - {}".format(bats, shs))
+
+    if bats:
+        command = " && ".join(bats)
+        return "{} && {}".format(command, cmd)
+    elif shs:
+        command = " && ".join(". ./{}".format(f) for f in shs)
+        return "{} && {}".format(command, cmd)
+    else:
+        return cmd
+
+
+class Environment:
+    def __init__(self):
+        # TODO: Maybe we need to pass conanfile to get the [conf]
+        # It being ordered allows for Windows case-insensitive composition
+        self._values = OrderedDict()  # {var_name: [] of values, including separators}
+
+    def __bool__(self):
+        return bool(self._values)
+
+    __nonzero__ = __bool__
+
+    def __repr__(self):
+        return repr(self._values)
+
+    def vars(self):
+        return list(self._values.keys())
+
+    def value(self, name, placeholder="{name}", pathsep=os.pathsep):
+        return self._format_value(name, self._values[name], placeholder, pathsep)
+
+    @staticmethod
+    def _format_value(name, varvalues, placeholder, pathsep):
+        values = []
+        for v in varvalues:
+
+            if v is _EnvVarPlaceHolder:
+                values.append(placeholder.format(name=name))
+            elif v is _PathSep:
+                values.append(pathsep)
+            else:
+                values.append(v)
+        return "".join(values)
+
+    @staticmethod
+    def _list_value(value, separator):
+        if isinstance(value, list):
```

Comment:
same

Reply:
I don't want to accept several different types of inputs, it should be either a value or a list. Passing an iterable, like a set, that is not ordered could produce unexpected results. Lets keep the interface as narrow as possible, there is always time to extend if we want.